### PR TITLE
fix: pornhub extractor

### PR DESCRIPTION
**Pull request by axetroy**
截屏2022-02-09 01 19 36

### CODE REVIEW

File: extractors/pornhub/pornhub.go [hunk 1]
```diff
@@ -24,11 +34,40 @@ func New() types.Extractor {
 
 // Extract is the main function to extract the data.
 func (e *extractor) Extract(url string, option types.Options) ([]*types.Data, error) {
-	html, err := request.Get(url, url, nil)
+	res, err := request.Request(http.MethodGet, url, nil, nil)
+
+	if err != nil {
```
Comment (iawia002) [inline extractors/pornhub/pornhub.go:39]:
```suggestion
	res, err := request.Request(http.MethodGet, url, nil, nil)
	if err != nil {
```

File: extractors/pornhub/pornhub.go [hunk 2]
```diff
@@ -24,11 +34,40 @@ func New() types.Extractor {
...
+		reader, _ = gzip.NewReader(res.Body)
+	case "deflate":
+		reader = flate.NewReader(res.Body)
+	default:
+		reader = res.Body
+	}
+	defer reader.Close() // nolint
+
+	b, err := ioutil.ReadAll(reader)
+
```
Comment (iawia002) [inline extractors/pornhub/pornhub.go:57]:
```suggestion
	b, err := ioutil.ReadAll(reader)
```

File: extractors/pornhub/pornhub.go [hunk 3]
```diff
@@ -37,46 +76,96 @@ func (e *extractor) Extract(url string, option types.Options) ([]*types.Data, er
 		title = "pornhub video"
 	}
 
-	realURLs := utils.MatchOneOf(html, `qualityItems_\d+\s=\s([^;]+)`)
-	if realURLs == nil || len(realURLs) < 2 {
-		return nil, types.ErrURLParseFailed
+	reg, err := regexp.Compile(`<script\b[^>]*>([\s\S]*?)<\/script>`)
+
+	if err != nil {
```
Comment (iawia002) [inline extractors/pornhub/pornhub.go:81]:
```suggestion
	reg, err := regexp.Compile(`<script\b[^>]*>([\s\S]*?)<\/script>`)
	if err != nil {
```

File: extractors/pornhub/pornhub.go [hunk 4]
```diff
@@ -37,46 +76,96 @@ func (e *extractor) Extract(url string, option types.Options) ([]*types.Data, er
...
+
+	if err := json.Unmarshal([]byte(resApi), &pornhubs); err != nil {
+		return nil, err
+	}
+
+	streams := make(map[string]*types.Stream, len(pornhubs))
+
+	for _, data := range pornhubs {
 		size, err := request.Size(data.VideoURL, data.VideoURL)
+
```
Comment (iawia002) [inline extractors/pornhub/pornhub.go:154]:
```suggestion
```

File: extractors/pornhub/pornhub.go [hunk 5]
```diff
@@ -37,46 +76,96 @@ func (e *extractor) Extract(url string, option types.Options) ([]*types.Data, er
...
+	}
+
+	flashId := regexp.MustCompile(`flashvars_\d+`).FindString(encryptedScript)
+
+	vm := otto.New()
+	vm.Run(`var playerObjList = {};` + encryptedScript + fmt.Sprintf(`;var __VM__OUTPUT = JSON.stringify(%s.mediaDefinitions)`, flashId))
+
+	value, err := vm.Get("__VM__OUTPUT")
+
+	if err != nil {
```
Comment (iawia002) [inline extractors/pornhub/pornhub.go:105]:
```suggestion
	value, err := vm.Get("__VM__OUTPUT")
	if err != nil {
```

File: extractors/pornhub/pornhub.go [hunk 6]
```diff
@@ -37,46 +76,96 @@ func (e *extractor) Extract(url string, option types.Options) ([]*types.Data, er
...
 
+	if mp4MediaDefinition == nil {
+		return nil, errors.New("can not found media")
+	}
+
+	resApi, err := request.Get(mp4MediaDefinition.VideoURL, mp4MediaDefinition.VideoURL, map[string]string{
+		"Cookie": strings.Join(cookiesArr, "; "),
+	})
+
+	if err != nil {
```
Comment (iawia002) [inline extractors/pornhub/pornhub.go:140]:
```suggestion
	resApi, err := request.Get(mp4MediaDefinition.VideoURL, mp4MediaDefinition.VideoURL, map[string]string{
		"Cookie": strings.Join(cookiesArr, "; "),
	})
	if err != nil {
```

File: extractors/pornhub/pornhub.go [hunk 7]
```diff
@@ -1,18 +1,28 @@
...
+	"io"
+	"io/ioutil"
+	"net/http"
+	"regexp"
 	"strings"
 
 	"github.com/iawia002/lux/extractors/types"
 	"github.com/iawia002/lux/request"
 	"github.com/iawia002/lux/utils"
+	"github.com/robertkrimen/otto"
```
Comment (iawia002) [inline extractors/pornhub/pornhub.go:18]: This should be in a separate group above

**Review by iawia002 — COMMENTED** [review]
Basically LGTM, only a few comments

File: extractors/pornhub/pornhub.go (hunk 7 above)
Comment (axetroy) [reply extractors/pornhub/pornhub.go:18]: This is gofmt's default behavior, this ordering is not up to me, I recommend keeping this

**Review by iawia002 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**